### PR TITLE
[WIP] Add S4 decoder in ESPnet2

### PR DESCRIPTION
**Pull request by m-koichi**
Hi, this PR adds an implementation of the S4 decoder (https://arxiv.org/abs/2210.17098) and a sample config.
The S4 codebase follows the official repository (https://github.com/HazyResearch/state-spaces), so the network configuration format is somewhat different from the models supported in ESPnet (e.g., Transformer decoder).

- LibriSpeech 960h WER result using this PR

|                     | dev_clean | dev_other | test_clean | test_other |
|:-----------:|----------:|-----------:|-----------:|-----------:|
| without LM | 2.0       | 5.0       | 2.3        | 5.0       |
| with LM       | 1.7       | 4.0       | 1.9        | 4.1        |

### CODE REVIEW

**Comment by sw005320** [comment]
Many thanks!
@pyf98, can you review this pr?

File: egs2/librispeech/asr1/conf/tuning/train_asr_s4_decoder.yaml [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    macaron_style: true
+    rel_pos_type: latest
+    pos_enc_layer_type: rel_pos
+    selfattention_layer_type: rel_selfattn
+    activation_type: swish
+    use_cnn_module: true
+    cnn_module_kernel: 31
+
+decoder: s4
+decoder_conf:
```
Comment (sw005320) [inline egs2/librispeech/asr1/conf/tuning/train_asr_s4_decoder.yaml:22]: Can you add some comments that the network configuration format is different?

File: espnet/nets/pytorch_backend/state_spaces/attention.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+import math
+
+import numpy
+import torch
+from torch import nn
+
+from espnet.nets.pytorch_backend.state_spaces.base import SequenceModule
+
+
+class MultiHeadedAttention(SequenceModule):
```
Comment (sw005320) [inline espnet/nets/pytorch_backend/state_spaces/attention.py:12]: Can you add a comment about why you need this instead of using our default MHA?

**Review by sw005320 — COMMENTED** [review]
- We may move the state_spaces directory under `espnet2/asr/` (if this is not easy or there is another reason, of course, we can keep it here).
- Can you upload a model checkpoint to the HF hub and add a model link and its result to `librispeech/asr1/REAME.md`?
- Can you also add a test? We can consider it later.

**Comment by m-koichi** [comment]
Sorry for the late reply.
> We may move the state_spaces directory under espnet2/asr/ (if this is not easy or there is another reason, of course, we can keep it here).

Since we tested S4 decoder in ASR and TTS tasks, we placed state_spaces directory under espnet/nets/pytorch_backend. Now I have moved the repository under espnet2/asr/
> Can you upload a model checkpoint to the HF hub and add a model link and its result to librispeech/asr1/REAME.md?

I have uploaded a model checkpoint to the HF hub [[link](https://huggingface.co/espnet/kmiyazaki_librispeech_asr_s4_decoder)]. I will update model card README.md later.

> Can you also add a test? We can consider it later.

I have added a simple test.

File: egs2/librispeech/asr1/conf/tuning/train_asr_s4_decoder.yaml [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+patience: none
+init: none
+best_model_criterion:
+-   - valid
+    - acc
+    - max
+keep_nbest_models: 10
+
+optim: adamw
+exclude_weight_decay: true
```
Comment (pyf98) [inline egs2/librispeech/asr1/conf/tuning/train_asr_s4_decoder.yaml:74]: Just curious: Is AdamW better than Adam? Also, does excluding weight decay for bias etc improve the performance?

File: espnet2/tasks/abs_task.py [hunk 4]
```diff
@@ -817,6 +818,19 @@ class ArgumentDefaultsRawTextHelpFormatter(
...
+            type=str2bool,
+            default=False,
+            help="Exclude weight decay in optimizer for model bias, normalization, "
+            "or other special parameters",
+        )
+        group.add_argument(
+            "--exclude_weight_decay_conf",
+            action=NestedDictAction,
+            default=dict(),
+            help="The keyword arguments for configuring weight decay in optimizer",
```
Comment (pyf98) [inline espnet2/tasks/abs_task.py:832]: Can you explain a bit more about this new argument?

**Comment by pyf98** [comment]
Overall, it looks good to me.

**Comment by sw005320** [comment]
Thanks, @pyf98 

@m-koichi, after you fix the CI issue and reflect @pyf98's comments, I'll merge this PR.

**Comment by sw005320** [comment]
We still have some CI errors.

File: egs2/librispeech/asr1/conf/tuning/train_asr_s4_decoder.yaml (hunk 3 above)
Comment (m-koichi) [reply egs2/librispeech/asr1/conf/tuning/train_asr_s4_decoder.yaml:74]: In our experiments, in S4 decoder case, we confirmed that AdamW is slightly better than Adam optimizer.
While in Transformer decoder case, there was no significant difference. 
For the weight decay trick, we followed the experimental setup of S4. Perhaps the effect may not be so apparent.

- S4 results (w/o LM)

|                     | dev_clean | dev_other | test_clean | test_other |
|---------------------|-----------|-----------|------------|------------|
| Adam                | 2.1       | 5.1       | 2.2        | 5.2        |
| AdamW               | 2.0       | 5.0       | 2.3        | 5.0        |
| AdamW w/o excluding | 2.0       | 5.0       | 2.2        | 5.0        |

Comment (pyf98) [reply egs2/librispeech/asr1/conf/tuning/train_asr_s4_decoder.yaml:74]: Thanks!

File: espnet2/tasks/abs_task.py (hunk 4 above)
Comment (m-koichi) [reply espnet2/tasks/abs_task.py]: Updated. Could you check the comment again?

File: egs2/librispeech/asr1/conf/tuning/train_asr_s4_decoder.yaml (hunk 3 above)
Comment (b-flo) [reply egs2/librispeech/asr1/conf/tuning/train_asr_s4_decoder.yaml:74]: Kind of out of scope but did someone experiment with the `ScaledAdam` from k2 team?

**Comment by sw005320** [comment]
https://github.com/espnet/espnet/actions/runs/3881628865/jobs/6637706128#step:8:18610
I think you can skip the test for the lower versions.
You can find some examples (e.g., https://github.com/espnet/espnet/blob/master/test/espnet2/enh/layers/test_complex_utils.py)

**Comment by m-koichi** [comment]
Thanks for the advice.
I fixed the test to skip for the lower pytorch versions.

**Comment by sw005320** [comment]
Many many thanks!
This is a really valuable contribution!

Could you add this new feature to https://github.com/espnet/espnet/blob/master/README.md to advertise this great function?

**Comment by m-koichi** [comment]
Thank you for reviewing the code and helping some CI issues.
All comments received have been reflected.
I updated HF hub model card to reproduce the result.
I added a bullet to the README related to the decoders of the ASR architecture.
`- Decoder: RNN (LSTM/GRU), Transformer, or S4`
Can I continue to commit to this branch? or should I create another pull request?

**Comment by sw005320** [comment]
You can create another PR.

**Comment by ftshijt** [comment]
Thanks for the contribution! 

I found that S4decoder needs an additional `hydra` package imported in `espnet2/asr/state_spaces/utils.py`. After my merge to this PR, it throws an error about `No module named 'hydra'` where I need to pip install. Not sure if we need to either check the import as an external package (as in the installer or makefile) or add that package to the `setup.py`
